### DISCUSSION TITLE
Event hook: Retrieve nested (expanded) JSON values

### DISCUSSION THREAD

**Discussion by YoRHa-A5** [pocketbase/pocketbase#2705] [Q&A] [answered]
- users location (latitude/longitude) is stored in a different table to ensure privacy
- users can enter a maxDistance value in their profile
- users want to be able to see other users within their given maxDistance (similar to dating apps)

To do this, I tried experimenting with an event hook:
```go
app.OnRecordsListRequest("users").Add(func(e *core.RecordsListEvent) error {
	data := map[string]any{
		"items": e.Result.Items,
	}

	apis.EnrichRecords(e.HttpContext, app.Dao(), e.Records)

	type Location struct {
		Latitude  float32 `json:"latitude"`
		Longitude float32 `json:"longitude"`
	}

	// result := struct {
	// 	location Location `json:"location"`
	// }{}

	for _, r := range e.Records {
		locationValue, ok := r.Expand()["location"]
		// log.Println(location.longitude)
		// log.Println(r.Expand()["location"])
		// r.UnmarshalJSONField("expand", &result)
		// log.Println(result)
	}
	
	e.HttpContext.JSON(http.StatusOK, data)
	return hook.StopPropagation
})
```

In the end using the  JS  SDK I want to be able to make a request, include the maxDistance value in the request, then inside the hook use a calculateDistance function that could result in a boolean depending on if the distance is within the user's maxDistance and using a conditional the response should only include the data of the users that fall within this condition.

A request with `?expand=location` results in the location key output like this:

```json
      "expand": {
        "location": {
          "collectionId": "xyz",
          "collectionName": "location",
          "created": "2023-06-12 15:35:44.592Z",
          "id": "xyz",
          "latitude": 12.34,
          "longitude": 56.78,
          "updated": "2023-06-12 15:35:44.592Z",
          "user": "xyz"
        }
      },
```

**Comment by ganigeorgiev** [comment, marked as answer]
I think the `OnRecordsListRequest` hook may not be suitable for your use case because you are breaking the pagination (you also don't know the total matching users this way). What if only 10 users of all others are "near by"?
I'm also confused by the `expand` result. Is `location` a direct relation in the `users` collection or it is a indirect relation? If it is the first, why there is a `user` field?

In any case here is an example with some guides if you still want to use the hook:

```go
app.OnRecordsListRequest("users").Add(func(e *core.RecordsListEvent) error {
	requestData := apis.RequestData(e.HttpContext)
	if requestData.Admin != nil || requestData.AuthRecord == nil {
		return nil // admin or guest -> proceed as usual
	}

	currentLocation, err := app.Dao().FindFirstRecordByData("locations", "user", requestData.AuthRecord.Id)
	if err != nil {
		return apis.NewBadRequestError("The current user doesn't have a location set.", err)
	}

	// We don't rely on expand because one user may not have access to the other users location
	// but if you want to access the expand record, you can do it with type assertion:
	// location := r.Expand()["location"].(*models.Record)
	//
	// Instead, we'll fetch each location separately and if the user is
	// "near by" will add it to the nearByUsers slice.
	// Optionally you can eager load the record ids and fetch the location at once,
	// but SQLite is performant enough so most of the times you don't have to worry about the N+1 problem
	// (see https://www.sqlite.org/np1queryprob.html)

	nearByUsers := make([]*models.Record, 0, len(e.Records))

	for _, r := range e.Records {
		// side-note: you may want to add an index on the "user" field if you don't have already
		location, err := app.Dao().FindFirstRecordByData("locations", "user", r.Id)
		if err != nil {
			continue // no location, skip
		}

		// pass the currentLocation and the expanded location fields to your calculate function
		isNearBy := calculateDistance(
			currentLocation.GetFloat("latitude"),
			currentLocation.GetFloat("longitude"),
			location.GetFloat("latitude"),
			location.GetFloat("longitude"),
		)

		// remove the user enty if it is not nearby
		if isNearBy {
			nearByUsers = append(nearByUsers, r)
		}
	}

	// overwrite event items, but what about the counts?
	e.Records = nearByUsers
	e.Result.Items = e.Records

	return nil
})
```

---

For your use case it may be better to create a custom route and perform the check there, something like (you may wan to add limits or):

```go
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
	e.Router.GET("/near-by-users", func(c echo.Context) error {
		requestData := apis.RequestData(c)

		locations, err := app.Dao().FindCollectionByNameOrId("locations")
		if err != nil {
			return apis.NewNotFoundError("locations colleciton is missing", err)
		}

		currentLocation, err := app.Dao().FindFirstRecordByData("locations", "user", requestData.AuthRecord.Id)
		if err != nil {
			return apis.NewBadRequestError("The current user doesn't have a location set.", err)
		}

		// find near by locations and then fetch their users
		// (adjust according to how calculateDistance works; you may need cos/sin conversion)
		nearByLocations := []*models.Record{}
		locErr := app.Dao().RecordQuery(locations).
			AndWhere(dbx.Not(dbx.HashExp{"id": currentLocation.Id})).
			AndWhere(dbx.NewExp("latitude >= {:fromLatitude}", dbx.Params{"fromLatitude": currentLocation.GetFloat("latitude") - 10})).     // some threshold
			AndWhere(dbx.NewExp("latitude <= {:toLatitude}", dbx.Params{"toLatitude": currentLocation.GetFloat("latitude") + 10})).         // some threshold
			AndWhere(dbx.NewExp("longitude >= {:fromLongitude}", dbx.Params{"fromLongitude": currentLocation.GetFloat("longitude") - 10})). // some threshold
			AndWhere(dbx.NewExp("longitude <= {:toLongitude}", dbx.Params{"toLongitude": currentLocation.GetFloat("longitude") + 10})).     // some threshold
			Limit(100).
			All(&nearByLocations)
		if locErr != nil {
			return apis.NewBadRequestError("Failed to fetch near by nearByLocations.", locErr)
		}

		if len(nearByLocations) == 0 {
			return c.JSON(http.StatusOK, []any{})
		}

		// fetch users with the nearByLocations
		userIds := make([]any, len(nearByLocations))
		for i, l := range nearByLocations {
			userIds[i] = l.Get("user")
		}

		nearByUsers, err := app.Dao().FindRecordsByExpr("users", dbx.HashExp{"id": userIds})
		if err != nil {
			return apis.NewBadRequestError("Failed to fetch near by users.", err)
		}

		return c.JSON(http.StatusOK, nearByUsers)
	}, apis.ActivityLogger(app), apis.RequireRecordAuth())

	return nil
})
```

**Reply by YoRHa-A5** [reply]
Thank you for the comprehensive reply. I will experiment with this 👍🏻 
Regarding the confusion: I think this stems from adding an (unnecessary) user relational field inside the location collection, which would be like a two way street instead of a one way relational connection starting from the user collection. Correct me if I'm wrong but it should be sufficient to have a relational field in the users collection linking to entries in the locations collection only containing lat/long each, no additional info needed, no?

**Reply by ganigeorgiev** [reply]
> Correct me if I'm wrong but it should be sufficient to have a relational field in the users collection linking to entries in the locations collection only containing lat/long each, no additional info needed, no?

Both options are fine (_if it is indirect relation then it is maybe slightly more clear that a single `locations` entry can be assigned to only 1 user_), but I'd recommend to avoid having the relation in 2 places as it can easily get out of sync.

**Reply by YoRHa-A5** [reply]
Ah sorry, deleted the question, since I failed to realize that it was another query, my code changed in the meantime :D However, this query and returning the data in the response results in a panic and I am not sure why:

```go
nearByUserRecords := []*models.Record{}
userErr := app.Dao().DB().Select("u.*", "p.birthday").
	From("users u").
	LeftJoin("users_private p", dbx.NewExp("u.id=p.user_id")).
	AndWhere(dbx.NewExp("p.latitude BETWEEN {:minLat} AND {:maxLat}", dbx.Params{"minLat": minLat, "maxLat": maxLat})).
	AndWhere(dbx.NewExp("p.longitude BETWEEN {:minLon} AND {:maxLon}", dbx.Params{"minLon": minLon, "maxLon": maxLon})).
	AndWhere(dbx.NewExp("u.id != {:activeUserId}", dbx.Params{"activeUserId": activeUserId})).
	All(&nearByUserRecords)

return c.JSON(http.StatusOK, nearByUserRecords)
```

**Reply by ganigeorgiev** [reply]
`models.Record` can be directly populated only if you use `RecordQuery()`.

If you want to construct a custom select data with combined fields from different tables, you can define a custom struct that describes the shape of the returned data.

**Reply by YoRHa-A5** [reply]
Maybe I am overcomplicating things here. Since the query happens on the server and I can then decide what info gets sent to the client, it should be safe to include private data in the users collection, no? But then I would need to be careful using the SDK user query functions, don't I?

Alternatively "hiding" the data inside expand would only work if I implement the private data record id inside the users collection!? 🤔
